File: 3AVAL/P02.py
import math

def encontraReta(x1, y1, x2, y2, x3, y3, A, B, C, Area, soma):
    theta = math.acos((C*C - B*B - A*A)/(-2.0*A*B))
    #prod = Area/math.sin(theta)
    prod  = A*B/2.0
    delta = soma*soma - 4*prod
    #print soma*soma, math.sin(theta), prod
    print delta
    #if delta < 0:
    #    delta *= -1
    
    r1 = (soma + math.sqrt(delta))/2.0
    r2 = (soma - math.sqrt(delta))/2.0

    s1 = soma - r1
    s2 = soma - r2

    #print "{0:.12f}".format(soma), "{0:.12f}".format(s1+r1), "{0:.12f}".format(s2+r2)

    if r1 <= B and s1 <= A:    
        p, q = x3 + (r1/B)* (x1 - x3), y3 + (r1/B) * (y1 - y3)
        k, l = x3 + (s1/A)* (x2 - x3), y3 + (s1/A) * (y2 - y3)
    elif r2 <= B and s2 <= A:
        p, q = x3 + (r2/B)* (x1 - x3), y3 + (r2/B) * (y1 - y3)
        k, l = x3 + (s2/A)* (x2 - x3), y3 + (s2/A) * (y2 - y3)
    else:
        return None

    return [p, q, k, l]


x1, y1, x2, y2, x3, y3 = map(float, raw_input().split())

#AB = (y1 - y2) * x + (x2 - x1) * y + x1 * y2 - x2 * y1
#AC = (y1 - y3) * x + (x3 - x1) * y + x1 * y3 - x3 * y1
#BC = (y2 - y3) * x + (x3 - x2) * y + x2 * y3 - x3 * y2
#reta = (yn - ybar) * x + (xbar - xn) * y + xn * ybar - xbar * yn

C = math.sqrt((x1-x2)*(x1-x2)+(y1-y2)*(y1-y2))
A = math.sqrt((x3-x2)*(x3-x2)+(y3-y2)*(y3-y2))
B = math.sqrt((x1-x3)*(x1-x3)+(y1-y3)*(y1-y3))

P = A + B + C

soma = P/2.0

Area = ((x3 - x1) * (x3 - x1) + (y3 - y1) * (y3 - y2))/2.0

#Area = math.sqrt(soma*(soma-A)*(soma-B)*(soma-C))



resp = encontraReta(x1, y1, x2, y2, x3, y3, A, B, C, Area, soma)
if resp == None:
    resp =  encontraReta(x2, y2, x3, y3, x1, y1, B, C, A, Area, soma)
if resp == None:
    resp =  encontraReta(x1, y1, x3, y3, x2, y2, A, C, B, Area, soma)
if resp == None:
    print "NO"
    exit(0)

    
print "YES"            
print "{0:.15f}".format(resp[0]), "{0:.15f}".format(resp[1])
print "{0:.15f}".format(resp[2]), "{0:.15f}".format(resp[3])

File: 3AVAL/ok/P02.cpp
/*
Nome: Shayenne Moura
NUSP: 8536235
MAC0327 - 2o sem 2015
3a aval
Problema P2 - 1647. Divide an Island!
 */
#include<iostream>
#include<math.h>
#include<stdio.h>

using namespace std;

int main() {
  double x1, y1, x2, y2, x3, y3;
  double A, B, C, tmp;
  double P, soma, prod, area, delta;
  int resp = 0;
  double r1, r2, s1, s2, p, q, k, l;
  
  cin >> x1 >> y1 >> x2 >> y2 >> x3 >> y3;

  A = sqrt ((x3-x2)*(x3-x2)+(y3-y2)*(y3-y2));
  B = sqrt ((x3-x1)*(x3-x1)+(y3-y1)*(y3-y1));
  C = sqrt ((x1-x2)*(x1-x2)+(y1-y2)*(y1-y2));

  P = A + B + C;

  soma = P/2;

  //area = ((x3 - x1) * (x3 - x1) + (y3 - y1) * (y3 - y2))/2;
  area = sqrt (soma*(soma-A)*(soma-B)*(soma-C));
  
  for (int i = 0; i < 3 && resp == 0; i++) {
    
    prod = A * B / 2;
    
    delta = soma*soma - 4*prod;
    if (fabs(delta) <= 1e-10)
      delta = 0;
    
    if (delta >= 0) {
      r1 = (soma + sqrt(delta))/2;
      r2 = (soma - sqrt(delta))/2;

      s1 = soma - r1;
      s2 = soma - r2;
      
      if (r1 <= B+1e-10 && s1 <= A+1e-10) {
	p = x3 + (r1/B) * (x1-x3);
	q = y3 + (r1/B) * (y1-y3);
	k = x3 + (s1/A) * (x2-x3);
	l = y3 + (s1/A) * (y2-y3);
	resp = 1;
      }
      else if (r2 <= B+1e-10 && s2 <= A+1e-10) {
	p = x3 + (r2/B) * (x1-x3);
	q = y3 + (r2/B) * (y1-y3);
	k = x3 + (s2/A) * (x2-x3);
	l = y3 + (s2/A) * (y2-y3);
	resp = 1;
      }
      
    }
    tmp = A;
    A = B;
    B = C;
    C = tmp;
    tmp = x1;
    x1 = x2;
    x2 = x3;
    x3 = tmp;
    tmp = y1;
    y1 = y2;
    y2 = y3;
    y3 = tmp;
  }

  if (! resp) {
    cout << "NO";
    return 0;
  }

  printf("YES\n%.15lf %.15lf\n%.15lf %.15lf\n", p, q, k, l);

  return 0;
}
